# Data Storytelling Project: SAT New York

**Objectives:**
* Analyse SAT scores + demographics and any other relevant information available
* Check fairness of the [SAT](https://www.insidehighered.com/news/2010/06/21/sat)


In [1]:
import pandas as pd
import numpy as np

## Understanding the data

### Read data into dataframe, put each dataframe into a dictionary

In [2]:
files = ["AP_results.csv", "Class_size.csv", "Demographics.csv", "Graduation_outcomes.csv", "hsdirectory.csv", "Math_test_results.csv", "SAT_results.csv"]
data = {}
for f in files:
    print f
    d = pd.read_csv("data/{}".format(f))
    data[f.replace(".csv", "")] = d

AP_results.csv
Class_size.csv
Demographics.csv
Graduation_outcomes.csv
hsdirectory.csv
Math_test_results.csv
SAT_results.csv


### Check dictionary

In [3]:
for k,v in data.items():
    print("\n" + k + "\n")
    print(v.head())


Math_test_results

      DBN Grade  Year      Category  Number Tested Mean Scale Score Level 1 #  \
0  01M015     3  2006  All Students             39              667         2   
1  01M015     3  2007  All Students             31              672         2   
2  01M015     3  2008  All Students             37              668         0   
3  01M015     3  2009  All Students             33              668         0   
4  01M015     3  2010  All Students             26              677         6   

  Level 1 % Level 2 # Level 2 % Level 3 # Level 3 % Level 4 # Level 4 %  \
0      5.1%        11     28.2%        20     51.3%         6     15.4%   
1      6.5%         3      9.7%        22       71%         4     12.9%   
2        0%         6     16.2%        29     78.4%         2      5.4%   
3        0%         4     12.1%        28     84.8%         1        3%   
4     23.1%        12     46.2%         6     23.1%         2      7.7%   

  Level 3+4 # Level 3+4 %  
0          26 

**Patterns/preliminary analysis:**

* District Borough Number (DBN): Math_test_results, SAT_results, Demographics, AP_results, Graduation_outcomes, hsdirectory
* School name: SAT_results, AP_results, Graduation_outcomes, Class_size, hsdirectory
* Location1 field in hsdirectory can be used for maps

### Data unification  

This makes working with all the data an easier process.
DBN is a common column, it just doesnt appear in Class_size - unless School code is DBN.

I will check how DBN looks and compare against Class_size to try to recognize a pattern

In [4]:
data["AP_results"]["DBN"].head()

0    01M448
1    01M450
2    01M515
3    01M539
4    02M296
Name: DBN, dtype: object

In [5]:
data["Class_size"].head()

CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   
3    1       M        M015  P.S. 015 Roberto Clemente     01          CTT   
4    1       M        M015  P.S. 015 Roberto Clemente     02       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   
3                                    -                                   -   
4                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   
3                           -                               17.0   
4                           -                               15.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   
3                 1.0                17.0                    17.0   
4                 1.0                15.0                    15.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  
0                   19.0         ATS                             NaN  
1                   21.0         ATS                             NaN  
2                   17.0         ATS                             NaN  
3                   17.0         ATS                             NaN  
4                   15.0         ATS                             NaN

It looks like DBN is combination of CSD, BOROUGH, and SCHOOL CODE.

from [string format](https://docs.python.org/2/library/stdtypes.html#str.format) [documentation](https://docs.python.org/3/library/string.html#format-string-syntax):

{:02d}

02d formats an integer (d) to a field of minimum width 2 (2), with zero-padding on the left (leading 0)


In [6]:
data["Class_size"]["DBN"] = data["Class_size"].apply(lambda x: "{0:02d}{1}".format(x["CSD"], x["SCHOOL CODE"]), axis=1)
data["hsdirectory"].rename(columns={'dbn': 'DBN'}, inplace=True)

In [7]:
data["Class_size"].head(3)

CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO     DBN  
0                   19.0         ATS                             NaN  01M015  
1                   21.0         ATS                             NaN  01M015  
2                   17.0         ATS                             NaN  01M015

I will now combine the survey data into one:

In [8]:
survey1 = pd.read_csv("./data/survey/masterfile11_gened_final.txt", delimiter="\t")
survey2 = pd.read_csv("./data/survey/masterfile11_d75_final.txt", delimiter="\t")

In [9]:
survey1.head(3)

dbn    bn                 schoolname  d75 studentssurveyed  highschool  \
0  01M015  M015  P.S. 015 Roberto Clemente    0               No         0.0   
1  01M019  M019        P.S. 019 Asher Levy    0               No         0.0   
2  01M020  M020       P.S. 020 Anna Silver    0               No         0.0   

          schooltype  rr_s  rr_t  rr_p     ...      s_N_q14e_3  s_N_q14e_4  \
0  Elementary School   NaN    88    60     ...             NaN         NaN   
1  Elementary School   NaN   100    60     ...             NaN         NaN   
2  Elementary School   NaN    88    73     ...             NaN         NaN   

   s_N_q14f_1  s_N_q14f_2  s_N_q14f_3  s_N_q14f_4  s_N_q14g_1  s_N_q14g_2  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   

   s_N_q14g_3  s_N_q14g_4  
0         NaN         NaN  
1         NaN         NaN  
2         NaN         NaN  

[3 rows x 1942 columns]

In [10]:
survey2.head(3)

dbn    bn schoolname  d75 studentssurveyed  highschool  \
0  75K004  K004  P.S. K004    1              Yes         0.0   
1  75K036  K036    P.S. 36    1              Yes         NaN   
2  75K053  K053  P.S. K053    1              Yes         NaN   

                      schooltype  rr_s  rr_t  rr_p    ...     s_q14_2  \
0  District 75 Special Education  38.0    90    72    ...        29.0   
1  District 75 Special Education  70.0    69    44    ...        20.0   
2  District 75 Special Education  94.0    97    53    ...        14.0   

   s_q14_3  s_q14_4  s_q14_5  s_q14_6  s_q14_7  s_q14_8  s_q14_9  s_q14_10  \
0      0.0      0.0      0.0      0.0      0.0      0.0      0.0       0.0   
1     27.0     19.0      9.0      2.0      6.0      1.0      2.0       0.0   
2     12.0     12.0     10.0     21.0     13.0     11.0      2.0       0.0   

   s_q14_11  
0       0.0  
1       0.0  
2       0.0  

[3 rows x 1773 columns]

Add 'False' to the d75 field in survey one (which corresponds to all schools) and 'True' to survey2 (which corresponds to school district 75)

In [11]:
survey1["d75"] = False
survey2["d75"] = True

Concatenate survey1 and survey2 vertically

In [12]:
survey = pd.concat([survey1, survey2], axis=0)

In [13]:
survey.head()

N_p    N_s   N_t  aca_p_11  aca_s_11  aca_t_11  aca_tot_11    bn  \
0   90.0    NaN  22.0       7.8       NaN       7.9         7.9  M015   
1  161.0    NaN  34.0       7.8       NaN       9.1         8.4  M019   
2  367.0    NaN  42.0       8.6       NaN       7.5         8.0  M020   
3  151.0  145.0  29.0       8.5       7.4       7.8         7.9  M034   
4   90.0    NaN  23.0       7.9       NaN       8.1         8.0  M063   

   com_p_11  com_s_11   ...    t_q8c_1  t_q8c_2  t_q8c_3 t_q8c_4  t_q9  \
0       7.6       NaN   ...       29.0     67.0      5.0     0.0   NaN   
1       7.6       NaN   ...       74.0     21.0      6.0     0.0   NaN   
2       8.3       NaN   ...       33.0     35.0     20.0    13.0   NaN   
3       8.2       5.9   ...       21.0     45.0     28.0     7.0   NaN   
4       7.9       NaN   ...       59.0     36.0      5.0     0.0   NaN   

   t_q9_1  t_q9_2  t_q9_3  t_q9_4  t_q9_5  
0     5.0    14.0    52.0    24.0     5.0  
1     3.0     6.0     3.0    78.0     9.0  
2     3.0     5.0    16.0    70.0     5.0  
3     0.0    18.0    32.0    39.0    11.0  
4    10.0     5.0    10.0    60.0    15.0  

[5 rows x 2773 columns]

Check data dictionary file (Survey Data Dictionary.xls) to define which columns are relevant

In [14]:
survey["DBN"] = survey["dbn"]
survey_fields = ["DBN", "rr_s", "rr_t", "rr_p", "N_s", "N_t", "N_p", "saf_p_11", "com_p_11", "eng_p_11", "aca_p_11", "saf_t_11", "com_t_11", "eng_t_10", "aca_t_11", "saf_s_11", "com_s_11", "eng_s_11", "aca_s_11", "saf_tot_11", "com_tot_11", "eng_tot_11", "aca_tot_11",]

In [15]:
survey.head()

N_p    N_s   N_t  aca_p_11  aca_s_11  aca_t_11  aca_tot_11    bn  \
0   90.0    NaN  22.0       7.8       NaN       7.9         7.9  M015   
1  161.0    NaN  34.0       7.8       NaN       9.1         8.4  M019   
2  367.0    NaN  42.0       8.6       NaN       7.5         8.0  M020   
3  151.0  145.0  29.0       8.5       7.4       7.8         7.9  M034   
4   90.0    NaN  23.0       7.9       NaN       8.1         8.0  M063   

   com_p_11  com_s_11   ...    t_q8c_2  t_q8c_3  t_q8c_4 t_q9  t_q9_1  t_q9_2  \
0       7.6       NaN   ...       67.0      5.0      0.0  NaN     5.0    14.0   
1       7.6       NaN   ...       21.0      6.0      0.0  NaN     3.0     6.0   
2       8.3       NaN   ...       35.0     20.0     13.0  NaN     3.0     5.0   
3       8.2       5.9   ...       45.0     28.0      7.0  NaN     0.0    18.0   
4       7.9       NaN   ...       36.0      5.0      0.0  NaN    10.0     5.0   

   t_q9_3  t_q9_4  t_q9_5     DBN  
0    52.0    24.0     5.0  01M015  
1     3.0    78.0     9.0  01M019  
2    16.0    70.0     5.0  01M020  
3    32.0    39.0    11.0  01M034  
4    10.0    60.0    15.0  01M063  

[5 rows x 2774 columns]

In [16]:
# getting only the relevant fields
# note that '.loc' was used because it works on lables. If I was working on the positions in the index, I would use iloc
survey = survey.loc[:,survey_fields]

In [17]:
survey.head()

DBN  rr_s  rr_t  rr_p    N_s   N_t    N_p  saf_p_11  com_p_11  eng_p_11  \
0  01M015   NaN    88    60    NaN  22.0   90.0       8.5       7.6       7.5   
1  01M019   NaN   100    60    NaN  34.0  161.0       8.4       7.6       7.6   
2  01M020   NaN    88    73    NaN  42.0  367.0       8.9       8.3       8.3   
3  01M034  89.0    73    50  145.0  29.0  151.0       8.8       8.2       8.0   
4  01M063   NaN   100    60    NaN  23.0   90.0       8.7       7.9       8.1   

      ...      eng_t_10  aca_t_11  saf_s_11  com_s_11  eng_s_11  aca_s_11  \
0     ...           NaN       7.9       NaN       NaN       NaN       NaN   
1     ...           NaN       9.1       NaN       NaN       NaN       NaN   
2     ...           NaN       7.5       NaN       NaN       NaN       NaN   
3     ...           NaN       7.8       6.2       5.9       6.5       7.4   
4     ...           NaN       8.1       NaN       NaN       NaN       NaN   

   saf_tot_11  com_tot_11  eng_tot_11  aca_tot_11  
0         8.0         7.7         7.5         7.9  
1         8.5         8.1         8.2         8.4  
2         8.2         7.3         7.5         8.0  
3         7.3         6.7         7.1         7.9  
4         8.5         7.6         7.9         8.0  

[5 rows x 23 columns]

I need to add the survey to the data dictionary

In [18]:
data["survey"] = survey
#survey.shape

In [19]:
data['survey'].head()

DBN  rr_s  rr_t  rr_p    N_s   N_t    N_p  saf_p_11  com_p_11  eng_p_11  \
0  01M015   NaN    88    60    NaN  22.0   90.0       8.5       7.6       7.5   
1  01M019   NaN   100    60    NaN  34.0  161.0       8.4       7.6       7.6   
2  01M020   NaN    88    73    NaN  42.0  367.0       8.9       8.3       8.3   
3  01M034  89.0    73    50  145.0  29.0  151.0       8.8       8.2       8.0   
4  01M063   NaN   100    60    NaN  23.0   90.0       8.7       7.9       8.1   

      ...      eng_t_10  aca_t_11  saf_s_11  com_s_11  eng_s_11  aca_s_11  \
0     ...           NaN       7.9       NaN       NaN       NaN       NaN   
1     ...           NaN       9.1       NaN       NaN       NaN       NaN   
2     ...           NaN       7.5       NaN       NaN       NaN       NaN   
3     ...           NaN       7.8       6.2       5.9       6.5       7.4   
4     ...           NaN       8.1       NaN       NaN       NaN       NaN   

   saf_tot_11  com_tot_11  eng_tot_11  aca_tot_11  
0         8.0         7.7         7.5         7.9  
1         8.5         8.1         8.2         8.4  
2         8.2         7.3         7.5         8.0  
3         7.3         6.7         7.1         7.9  
4         8.5         7.6         7.9         8.0  

[5 rows x 23 columns]

### Checking the datasets

In [20]:
for k,v in data.items():
    print("\n" + k + "\n")
    print(v.head(3))


Math_test_results

      DBN Grade  Year      Category  Number Tested Mean Scale Score Level 1 #  \
0  01M015     3  2006  All Students             39              667         2   
1  01M015     3  2007  All Students             31              672         2   
2  01M015     3  2008  All Students             37              668         0   

  Level 1 % Level 2 # Level 2 % Level 3 # Level 3 % Level 4 # Level 4 %  \
0      5.1%        11     28.2%        20     51.3%         6     15.4%   
1      6.5%         3      9.7%        22       71%         4     12.9%   
2        0%         6     16.2%        29     78.4%         2      5.4%   

  Level 3+4 # Level 3+4 %  
0          26       66.7%  
1          26       83.9%  
2          31       83.8%  

SAT_results

      DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNIT

Math_test_results, Demographics, Graduation_outcomes, Class_size have several rows for each high school (DBN and School name fields).

I will need to find a way to condensate the datasets above to "one row per high school". If that is not done, it won't be possible to compare the datasets.

** Class_size:**

In [21]:
data["Class_size"]

CSD BOROUGH SCHOOL CODE  \
0        1       M        M015   
1        1       M        M015   
2        1       M        M015   
3        1       M        M015   
4        1       M        M015   
5        1       M        M015   
6        1       M        M015   
7        1       M        M015   
8        1       M        M015   
9        1       M        M015   
10       1       M        M015   
11       1       M        M015   
12       1       M        M019   
13       1       M        M019   
14       1       M        M019   
15       1       M        M019   
16       1       M        M019   
17       1       M        M019   
18       1       M        M019   
19       1       M        M019   
20       1       M        M019   
21       1       M        M019   
22       1       M        M019   
23       1       M        M019   
24       1       M        M019   
25       1       M        M020   
26       1       M        M020   
27       1       M        M020   
28       1       M        M020   
29       1       M        M020   
...    ...     ...         ...   
27581   32       K        K556   
27582   32       K        K556   
27583   32       K        K556   
27584   32       K        K556   
27585   32       K        K556   
27586   32       K        K556   
27587   32       K        K556   
27588   32       K        K556   
27589   32       K        K556   
27590   32       K        K556   
27591   32       K        K556   
27592   32       K        K556   
27593   32       K        K556   
27594   32       K        K556   
27595   32       K        K564   
27596   32       K        K564   
27597   32       K        K564   
27598   32       K        K564   
27599   32       K        K564   
27600   32       K        K564   
27601   32       K        K564   
27602   32       K        K564   
27603   32       K        K564   
27604   32       K        K564   
27605   32       K        K564   
27606   32       K        K564   
27607   32       K        K564   
27608   32       K        K564   
27609   32       K        K564   
27610   32       K        K564   

                                             SCHOOL NAME GRADE  PROGRAM TYPE  \
0                              P.S. 015 Roberto Clemente     0K       GEN ED   
1                              P.S. 015 Roberto Clemente     0K          CTT   
2                              P.S. 015 Roberto Clemente     01       GEN ED   
3                              P.S. 015 Roberto Clemente     01          CTT   
4                              P.S. 015 Roberto Clemente     02       GEN ED   
5                              P.S. 015 Roberto Clemente     02          CTT   
6                              P.S. 015 Roberto Clemente     03       GEN ED   
7                              P.S. 015 Roberto Clemente     03          CTT   
8                              P.S. 015 Roberto Clemente     04       GEN ED   
9                              P.S. 015 Roberto Clemente     05       GEN ED   
10                             P.S. 015 Roberto Clemente  0K-09      SPEC ED   
11                             P.S. 015 Roberto Clemente    NaN          NaN   
12                                   P.S. 019 Asher Levy     0K       GEN ED   
13                                   P.S. 019 Asher Levy     0K          CTT   
14                                   P.S. 019 Asher Levy     01       GEN ED   
15                                   P.S. 019 Asher Levy     01          CTT   
16                                   P.S. 019 Asher Levy     02       GEN ED   
17                                   P.S. 019 Asher Levy     03       GEN ED   
18                                   P.S. 019 Asher Levy     04       GEN ED   
19                                   P.S. 019 Asher Levy     04          CTT   
20                                   P.S. 019 Asher Levy     05       GEN ED   
21                                   P.S. 019 Asher Levy     05          CTT   
22                                   P.S. 019 Asher Levy  

It looks like grade 9-12 corresponds to what I [want](https://www.justlanded.com/english/United-States/USA-Guide/Education/The-American-school-system).

From the dataset [description](https://data.cityofnewyork.us/Education/2010-2011-Class-Size-School-level-detail/urz7-pzb3), I've learnt that there are 3 program types:

"Average class sizes for each school, by grade and program type (General Education, Self-Contained Special Education, Collaborative Team Teaching (CTT)) for grades K-9 (where grade 9 is not reported by subject area), and for grades 5-9 (where available) and 9-12, aggregated by program type (General Education, CTT, and Self-Contained Special Education) and core course (e.g. English 9, Integrated Algebra, US History, etc.)."

I can focus on General Education for now.

To group the data, I will have to average the values. So I will group Class_size by DBN (so the data will be "one row per high school")  and average each column. I will get the average Class_size for each school


In [22]:
Class_size = data["Class_size"]
Class_size = Class_size[Class_size["GRADE "] == "09-12"]
Class_size = Class_size[Class_size["PROGRAM TYPE"] == "GEN ED"]
Class_size = Class_size.groupby("DBN").agg(np.mean)

In [23]:
Class_size.head()

CSD  NUMBER OF STUDENTS / SEATS FILLED  NUMBER OF SECTIONS  \
DBN                                                                  
01M292    1                            88.0000            4.000000   
01M332    1                            46.0000            2.000000   
01M378    1                            33.0000            1.000000   
01M448    1                           105.6875            4.750000   
01M450    1                            57.6000            2.733333   

        AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  SIZE OF LARGEST CLASS  \
DBN                                                                         
01M292           22.564286                   18.50              26.571429   
01M332           22.000000                   21.00              23.500000   
01M378           33.000000                   33.00              33.000000   
01M448           22.231250                   18.25              27.062500   
01M450           21.200000                   19.40              22.866667   

        SCHOOLWIDE PUPIL-TEACHER RATIO  
DBN                                     
01M292                             NaN  
01M332                             NaN  
01M378                             NaN  
01M448                             NaN  
01M450                             NaN

DBN is now the index. It needs to be a "normal" column

In [24]:
Class_size.reset_index(inplace=True)
data["Class_size"] = Class_size

In [25]:
data["Class_size"].head()

DBN  CSD  NUMBER OF STUDENTS / SEATS FILLED  NUMBER OF SECTIONS  \
0  01M292    1                            88.0000            4.000000   
1  01M332    1                            46.0000            2.000000   
2  01M378    1                            33.0000            1.000000   
3  01M448    1                           105.6875            4.750000   
4  01M450    1                            57.6000            2.733333   

   AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  SIZE OF LARGEST CLASS  \
0           22.564286                   18.50              26.571429   
1           22.000000                   21.00              23.500000   
2           33.000000                   33.00              33.000000   
3           22.231250                   18.25              27.062500   
4           21.200000                   19.40              22.866667   

   SCHOOLWIDE PUPIL-TEACHER RATIO  
0                             NaN  
1                             NaN  
2                             NaN  
3                             NaN  
4                             NaN

**Math_test_results:**

DBN has been entered by Grade and by Year, as data was collected for multiple years for the same schools

In [26]:
data["Math_test_results"]["Year"].max()

2011

The most recent year is 2011, which will be the one I will use then.

In [27]:
data["Math_test_results"] = data["Math_test_results"][data["Math_test_results"]["Year"] == 2011]

In [28]:
data["Math_test_results"].head()

DBN       Grade  Year      Category  Number Tested Mean Scale Score  \
5   01M015           3  2011  All Students             28              671   
11  01M015           4  2011  All Students             28              668   
17  01M015           5  2011  All Students             25              667   
24  01M015  All Grades  2011  All Students             81              669   
30  01M019           3  2011  All Students             34              679   

   Level 1 # Level 1 % Level 2 # Level 2 % Level 3 # Level 3 % Level 4 #  \
5         10     35.7%        13     46.4%         5     17.9%         0   
11         3     10.7%        14       50%         9     32.1%         2   
17         5       20%         8       32%        12       48%         0   
24        18     22.2%        35     43.2%        26     32.1%         2   
30         2      5.9%        23     67.6%         7     20.6%         2   

   Level 4 % Level 3+4 # Level 3+4 %  
5         0%           5       17.9%  
11      7.1%          11       39.3%  
17        0%          12         48%  
24      2.5%          28       34.6%  
30      5.9%           9       26.5%

I will also need to select only one Grade, so I will pick the max one which is 8

In [29]:
data["Math_test_results"]["Grade"].dtype

dtype('O')

This is an object type. From [stackoverflow](https://stackoverflow.com/questions/34881079/pandas-distinction-between-str-and-object-types):

"As a very brief explanation that isn't a full answer: If you use a string dtype in numpy, it's fundamentally a fixed-width c-like string. In pandas, they're "normal" python strings, thus the object type"

(That's why it was not working when in the code below I did  == 8)

In [30]:
data["Math_test_results"] = data["Math_test_results"][data["Math_test_results"]["Grade"] == "8"]

In [31]:
data["Math_test_results"].head()

DBN Grade  Year      Category  Number Tested Mean Scale Score  \
111  01M034     8  2011  All Students             48              646   
280  01M140     8  2011  All Students             61              665   
346  01M184     8  2011  All Students             49              727   
388  01M188     8  2011  All Students             49              658   
411  01M292     8  2011  All Students             49              650   

    Level 1 # Level 1 % Level 2 # Level 2 % Level 3 # Level 3 % Level 4 #  \
111        15     31.3%        22     45.8%        11     22.9%         0   
280         1      1.6%        43     70.5%        17     27.9%         0   
346         0        0%         0        0%         5     10.2%        44   
388        10     20.4%        26     53.1%        10     20.4%         3   
411        15     30.6%        25       51%         7     14.3%         2   

    Level 4 % Level 3+4 # Level 3+4 %  
111        0%          11       22.9%  
280        0%          17       27.9%  
346     89.8%          49        100%  
388      6.1%          13       26.5%  
411      4.1%           9       18.4%

**Demographics:**

In [32]:
data["Demographics"]

DBN                                              Name  schoolyear  \
0      01M015                         P.S. 015 ROBERTO CLEMENTE    20052006   
1      01M015                         P.S. 015 ROBERTO CLEMENTE    20062007   
2      01M015                         P.S. 015 ROBERTO CLEMENTE    20072008   
3      01M015                         P.S. 015 ROBERTO CLEMENTE    20082009   
4      01M015                         P.S. 015 ROBERTO CLEMENTE    20092010   
5      01M015                         P.S. 015 ROBERTO CLEMENTE    20102011   
6      01M015  P.S. 015 ROBERTO CLEMENTE                           20112012   
7      01M019                               P.S. 019 ASHER LEVY    20052006   
8      01M019                               P.S. 019 ASHER LEVY    20062007   
9      01M019                               P.S. 019 ASHER LEVY    20072008   
10     01M019                               P.S. 019 ASHER LEVY    20082009   
11     01M019                               P.S. 019 ASHER LEVY    20092010   
12     01M019                               P.S. 019 ASHER LEVY    20102011   
13     01M019  P.S. 019 ASHER LEVY                                 20112012   
14     01M020                                PS 020 ANNA SILVER    20052006   
15     01M020                                PS 020 ANNA SILVER    20062007   
16     01M020                                PS 020 ANNA SILVER    20072008   
17     01M020                                PS 020 ANNA SILVER    20082009   
18     01M020                                PS 020 ANNA SILVER    20092010   
19     01M020                                PS 020 ANNA SILVER    20102011   
20     01M020  PS 020 ANNA SILVER                                  20112012   
21     01M034                       PS 034 FRANKLIN D ROOSEVELT    20052006   
22     01M034                       PS 034 FRANKLIN D ROOSEVELT    20062007   
23     01M034                       PS 034 FRANKLIN D ROOSEVELT    20072008   
24     01M034                       PS 034 FRANKLIN D ROOSEVELT    20082009   
25     01M034                       PS 034 FRANKLIN D ROOSEVELT    20092010   
26     01M034                       PS 034 FRANKLIN D ROOSEVELT    20102011   
27     01M034  PS 034 FRANKLIN D ROOSEVELT                         20112012   
28     01M056                                  JHS 056 CORLEARS    20052006   
29     01M063                           PS 063 WILLIAM MCKINLEY    20052006   
...       ...                                               ...         ...   
10045  32K549                BUSHWICK SCHOOL FOR SOCIAL JUSTICE    20102011   
10046  32K549  BUSHWICK SCHOOL FOR SOCIAL JUSTICE                  20112012   
10047  32K552                         ACADEMY OF URBAN PLANNING    20052006   
10048  32K552                         ACADEMY OF URBAN PLANNING    20062007   
10049  32K552                         ACADEMY OF URBAN PLANNING    20072008   
10050  32K552                         ACADEMY OF URBAN PLANNING    20082009   
10051  32K552                         ACADEMY OF URBAN PLANNING    20092010   
10052  32K552                         ACADEMY OF URBAN PLANNING    20102011   
10053  32K552  ACADEMY OF URBAN PLANNING                           20112012   
10054  32K554              ALL CITY LEADERSHIP SECONDARY SCHOOL    20052006   
10055  32K554              ALL CITY LEADERSHIP SECONDARY SCHOOL    20062007   
10056  32K554              ALL CITY LEADERSHIP SECONDARY SCHOOL    20072008   
10057  32K554              ALL CITY LEADERSHIP SECONDARY SCHOOL    20082009   
10058  32K554              ALL CITY LEADERSHIP SECONDARY SCHOOL    20092010   
10059  32K554              ALL CITY LEADERSHIP SECONDARY SCHOOL    20102011   
10060  32K554  ALL CITY LEADERSHIP SECONDARY SCHOOL                20112012   
10061  32K556  BUSHWICK LEADERS HIGH SCHOOL FOR ACADEMIC EXCELL    20052006   
10062  32K556  BUSHWICK LEADERS HIGH SCHOOL FOR ACADEMIC EXCELL    20062007   
10063  32K556  BUSHWICK LEADERS HIGH SCHOOL FOR ACADEMIC EXC

Again data was collected for multiple years for the same school

In [33]:
data["Demographics"]["schoolyear"].dtype

dtype('int64')

In [34]:
data["Demographics"] = data["Demographics"][data["Demographics"]["schoolyear"] == 20112012]

In [35]:
data["Demographics"].head(3)

DBN                                              Name  schoolyear  \
6   01M015  P.S. 015 ROBERTO CLEMENTE                           20112012   
13  01M019  P.S. 019 ASHER LEVY                                 20112012   
20  01M020  PS 020 ANNA SILVER                                  20112012   

   fl_percent  frl_percent  total_enrollment prek    k grade1 grade2  \
6         NaN         89.4               189   13   31     35     28   
13        NaN         61.5               328   32   46     52     54   
20        NaN         92.5               626   52  102    121     87   

      ...     black_num black_per hispanic_num hispanic_per white_num  \
6     ...            63      33.3          109         57.7         4   
13    ...            81      24.7          158         48.2        28   
20    ...            55       8.8          357         57.0        16   

   white_per male_num male_per female_num female_per  
6        2.1     97.0     51.3       92.0       48.7  
13       8.5    147.0     44.8      181.0       55.2  
20       2.6    330.0     52.7      296.0       47.3  

[3 rows x 38 columns]

**Graduation_outcomes:**

In [36]:
data["Graduation_outcomes"]

Demographic     DBN                            School Name    Cohort  \
0      Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2003   
1      Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2004   
2      Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2005   
3      Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2006   
4      Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL  2006 Aug   
5      Total Cohort  01M448    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL      2001   
6      Total Cohort  01M448    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL      2002   
7      Total Cohort  01M448    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL      2003   
8      Total Cohort  01M448    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL      2004   
9      Total Cohort  01M448    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL      2005   
10     Total Cohort  01M448    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL      2006   
11     Total Cohort  01M448    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL  2006 Aug   
12     Total Cohort  01M450             EAST SIDE COMMUNITY SCHOOL      2001   
13     Total Cohort  01M450             EAST SIDE COMMUNITY SCHOOL      2002   
14     Total Cohort  01M450             EAST SIDE COMMUNITY SCHOOL      2003   
15     Total Cohort  01M450             EAST SIDE COMMUNITY SCHOOL      2004   
16     Total Cohort  01M450             EAST SIDE COMMUNITY SCHOOL      2005   
17     Total Cohort  01M450             EAST SIDE COMMUNITY SCHOOL      2006   
18     Total Cohort  01M450             EAST SIDE COMMUNITY SCHOOL  2006 Aug   
19     Total Cohort  01M509                MARTA VALLE HIGH SCHOOL      2001   
20     Total Cohort  01M509                MARTA VALLE HIGH SCHOOL      2002   
21     Total Cohort  01M509                MARTA VALLE HIGH SCHOOL      2003   
22     Total Cohort  01M509                MARTA VALLE HIGH SCHOOL      2004   
23     Total Cohort  01M509                MARTA VALLE HIGH SCHOOL      2005   
24     Total Cohort  01M509                MARTA VALLE HIGH SCHOOL      2006   
25     Total Cohort  01M509                MARTA VALLE HIGH SCHOOL  2006 Aug   
26     Total Cohort  01M515  LOWER EAST SIDE PREPARATORY HIGH SCHO      2001   
27     Total Cohort  01M515  LOWER EAST SIDE PREPARATORY HIGH SCHO      2002   
28     Total Cohort  01M515  LOWER EAST SIDE PREPARATORY HIGH SCHO      2003   
29     Total Cohort  01M515  LOWER EAST SIDE PREPARATORY HIGH SCHO      2004   
...             ...     ...                                    ...       ...   
25066          Male  32K549     BUSHWICK SCHOOL FOR SOCIAL JUSTICE      2003   
25067          Male  32K549     BUSHWICK SCHOOL FOR SOCIAL JUSTICE      2004   
25068          Male  32K549     BUSHWICK SCHOOL FOR SOCIAL JUSTICE      2005   
25069          Male  32K549     BUSHWICK SCHOOL FOR SOCIAL JUSTICE      2006   
25070          Male  32K549     BUSHWICK SCHOOL FOR SOCIAL JUSTICE  2006 Aug   
25071          Male  32K552              ACADEMY OF URBAN PLANNING      2002   
25072          Male  32K552              ACADEMY OF URBAN PLANNING      2003   
25073          Male  32K552              ACADEMY OF URBAN PLANNING      2004   
25074          Male  32K552              ACADEMY OF URBAN PLANNING      2005   
25075          Male  32K552              ACADEMY OF URBAN PLANNING      2006   
25076          Male  32K552              ACADEMY OF URBAN PLANNING  2006 Aug   
25077          Male  32K554   ALL CITY LEADERSHIP SECONDARY SCHOOL      2002   
25078          Male  32K554   ALL CITY LEADERSHIP SECONDARY SCHOOL      2003   
25079          Male  32K554   ALL CITY LEADERSHIP SECONDARY SCHOOL      2004   
25080          Male  32K554   ALL CITY LEADERSHIP SECONDARY SCHOOL      2005   
25081          Male  32K554   ALL CITY LEADERSHIP SECONDARY SCHOOL      2006   
25082          Male  32K554   ALL CITY LEADERSHIP SECONDARY SCHOOL  2006 Aug   
25083          Male  32K556  BUSHWICK LEADERS HIGH SCHOOL FOR ACAD      2002   
25084   

I need to select Cohort == 2006

In [37]:
(data["Graduation_outcomes"]['Cohort'] == '2006').any()

True

In [38]:
data["Graduation_outcomes"]['Cohort'].dtype

dtype('O')

In [39]:
data["Graduation_outcomes"] = data["Graduation_outcomes"][data["Graduation_outcomes"]["Cohort"] == "2006"]
data["Graduation_outcomes"] = data["Graduation_outcomes"][data["Graduation_outcomes"]["Demographic"] == "Total Cohort"]

Check data dictionary

In [40]:
for k,v in data.items():
    print("\n" + k + "\n")
    print(v.head())


Math_test_results

        DBN Grade  Year      Category  Number Tested Mean Scale Score  \
111  01M034     8  2011  All Students             48              646   
280  01M140     8  2011  All Students             61              665   
346  01M184     8  2011  All Students             49              727   
388  01M188     8  2011  All Students             49              658   
411  01M292     8  2011  All Students             49              650   

    Level 1 # Level 1 % Level 2 # Level 2 % Level 3 # Level 3 % Level 4 #  \
111        15     31.3%        22     45.8%        11     22.9%         0   
280         1      1.6%        43     70.5%        17     27.9%         0   
346         0        0%         0        0%         5     10.2%        44   
388        10     20.4%        26     53.1%        10     20.4%         3   
411        15     30.6%        25       51%         7     14.3%         2   

    Level 4 % Level 3+4 # Level 3+4 %  
111        0%          11       22.9% 

### Computations

**SAT score:**

I will compute the _Total_ SAT score (SAT Math Avg. Score + SAT Critical Reading Avg. Score + SAT Writing Avg. Score) so I can compare the schools

In [43]:
columns = ['SAT Math Avg. Score', 'SAT Critical Reading Avg. Score', 'SAT Writing Avg. Score']
for c in columns:
    data["SAT_results"][c] = data["SAT_results"][c].apply(pd.to_numeric, errors='ignore') #converting string to number

data['SAT_results']['SAT_score'] = data['SAT_results'][columns[0]] + data['SAT_results'][columns[1]] + data['SAT_results'][columns[2]]

In [44]:
data["SAT_results"]

DBN                                        SCHOOL NAME  \
0    01M292      HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1    01M448                UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2    01M450                         EAST SIDE COMMUNITY SCHOOL   
3    01M458                          FORSYTH SATELLITE ACADEMY   
4    01M509                            MARTA VALLE HIGH SCHOOL   
5    01M515            LOWER EAST SIDE PREPARATORY HIGH SCHOOL   
6    01M539  NEW EXPLORATIONS INTO SCIENCE, TECHNOLOGY AND ...   
7    01M650                               CASCADES HIGH SCHOOL   
8    01M696                     BARD HIGH SCHOOL EARLY COLLEGE   
9    02M047  47 THE AMERICAN SIGN LANGUAGE AND ENGLISH SECO...   
10   02M288                       FOOD AND FINANCE HIGH SCHOOL   
11   02M294                               ESSEX STREET ACADEMY   
12   02M296              HIGH SCHOOL OF HOSPITALITY MANAGEMENT   
13   02M298                                   PACE HIGH SCHOOL   
14   02M300  URBAN ASSEMBLY SCHOOL OF DESIGN AND CONSTRUCTI...   
15   02M303                         FACING HISTORY SCHOOL, THE   
16   02M305  URBAN ASSEMBLY ACADEMY OF GOVERNMENT AND LAW, THE   
17   02M308                       LOWER MANHATTAN ARTS ACADEMY   
18   02M313  JAMES BALDWIN SCHOOL, THE: A SCHOOL FOR EXPEDI...   
19   02M316  URBAN ASSEMBLY SCHOOL OF BUSINESS FOR YOUNG WO...   
20   02M374                          GRAMERCY ARTS HIGH SCHOOL   
21   02M376                                        NYC ISCHOOL   
22   02M392                         MANHATTAN BUSINESS ACADEMY   
23   02M393                          BUSINESS OF SPORTS SCHOOL   
24   02M394                           EMMA LAZARUS HIGH SCHOOL   
25   02M399         THE HIGH SCHOOL FOR LANGUAGE AND DIPLOMACY   
26   02M400              HIGH SCHOOL FOR ENVIRONMENTAL STUDIES   
27   02M407              INSTITUTE FOR COLLABORATIVE EDUCATION   
28   02M408           PROFESSIONAL PERFORMING ARTS HIGH SCHOOL   
29   02M411                  BARUCH COLLEGE CAMPUS HIGH SCHOOL   
..      ...                                                ...   
448  30Q580          BACCALAUREATE SCHOOL FOR GLOBAL EDUCATION   
449  31R047          CSI HIGH SCHOOL FOR INTERNATIONAL STUDIES   
450  31R064        GAYNOR MCCOWN EXPEDITIONARY LEARNING SCHOOL   
451  31R080                     THE MICHAEL J. PETRIDES SCHOOL   
452  31R440                               NEW DORP HIGH SCHOOL   
453  31R445                          PORT RICHMOND HIGH SCHOOL   
454  31R450                                 CURTIS HIGH SCHOOL   
455  31R455                            TOTTENVILLE HIGH SCHOOL   
456  31R460                        SUSAN E. WAGNER HIGH SCHOOL   
457  31R470                                CONCORD HIGH SCHOOL   
458  31R600  RALPH R. MCKEE CAREER AND TECHNICAL EDUCATION ...   
459  31R605                STATEN ISLAND TECHNICAL HIGH SCHOOL   
460  31R607    YOUNG ADULT BOROUGH CENTER AT STATEN IS TECH HS   
461  32K403               ACADEMY FOR ENVIRONMENTAL LEADERSHIP   
462  32K545        EBC HIGH SCHOOL FOR PUBLIC SERVICE–BUSHWICK   
463  32K549                 BUSHWICK SCHOOL FOR SOCIAL JUSTICE   
464  32K552                          ACADEMY OF URBAN PLANNING   
465  32K554               ALL CITY LEADERSHIP SECONDARY SCHOOL   
466  32K556  BUSHWICK LEADERS HIGH SCHOOL FOR ACADEMIC EXCE...   
467  32K564                     BUSHWICK COMMUNITY HIGH SCHOOL   
468  75K371                      P.S. 371 s LILLIAN L. RASHKIS   
469  75M035                                           P.S. 035   
470  75Q256                                          P.S. Q256   
471  75Q811                                          P.S. Q811   
472  75R025            SOUTH RICHMOND HIGH SCHOOL I.S./P.S. 25   
473  75X012                   P.S. X012 LEWIS AND CLARK SCHOOL   
474  75X754            J. M. RAPPORT SCHOOL CAREER DEVELOPMENT   
475  79M645         SCHOOL FOR COOPERATIVE TECHNICAL EDUCATION   
476  79Q950                                GED P

**Get schools coordinates for maps:**

hsdirectory has a field 'Location 1' with the lat and lon of each school

In [45]:
data["hsdirectory"]['Location 1'][0]

'883 Classon Avenue\nBrooklyn, NY 11225\n(40.67029890700047, -73.96164787599963)'

Step by step:
1. I am creating a lambda function that will split by the string "\n"
2. I will them have 3 parts: '883 Classon Avenue', 'Brooklyn, NY 11225' and '(40.67029890700047, -73.96164787599963)'
3. I know I want the last bit. I can access it by either using [2] or [-1], as it is the last part
4. Then I need to remove both () in (40.67029890700047, -73.96164787599963)
5. I am left with 40.67029890700047, -73.96164787599963; so I will [0] to get lat and [1] to get lon ([-1] would also work for lon).

In [66]:
data["hsdirectory"]['lat'] = data["hsdirectory"]['Location 1'].apply(lambda x: x.split("\n")[-1].replace("(", "").replace(")", "").split(", ")[0])
data["hsdirectory"]['lon'] = data["hsdirectory"]['Location 1'].apply(lambda x: x.split("\n")[-1].replace("(", "").replace(")", "").split(", ")[1])

In [67]:
data["hsdirectory"]

DBN                                        school_name           boro  \
0    17K548                Brooklyn School for Music & Theatre       Brooklyn   
1    09X543                   High School for Violin and Dance          Bronx   
2    09X327        Comprehensive Model School Project M.S. 327          Bronx   
3    02M280     Manhattan Early College School for Advertising      Manhattan   
4    28Q680  Queens Gateway to Health Sciences Secondary Sc...         Queens   
5    08X348              Schuylerville Preparatory High School          Bronx   
6    14K474      PROGRESS High School for Professional Careers       Brooklyn   
7    02M420  High School for Health Professions and Human S...      Manhattan   
8    28Q350                    Jamaica Gateway to the Sciences         Queens   
9    29Q313  Benjamin Franklin High School for Finance & In...         Queens   
10   12X271                  East Bronx Academy for the Future          Bronx   
11   02M520   Murry Bergtraum High School for Business Careers      Manhattan   
12   06M211  Inwood Early College for Health and Informatio...      Manhattan   
13   24Q600        Queens Vocational and Technical High School         Queens   
14   10X368           In-Tech Academy (M.S. / High School 368)          Bronx   
15   06M348   Washington Heights Expeditionary Learning School      Manhattan   
16   18K629         Cultural Academy for the Arts and Sciences       Brooklyn   
17   03M299      High School for Arts, Imagination and Inquiry      Manhattan   
18   07X520         Foreign Language Academy of Global Studies          Bronx   
19   08X312                             Millennium Art Academy          Bronx   
20   08X293  Renaissance High School for Musical Theater & ...          Bronx   
21   13K419  Science Skills Center High School for Science,...       Brooklyn   
22   31R460                        Susan E. Wagner High School  Staten Island   
23   28Q328               High School for Community Leadership         Queens   
24   05M362                          Columbia Secondary School      Manhattan   
25   10X243                  West Bronx Academy for the Future          Bronx   
26   03M307       Urban Assembly School for Media Studies, The      Manhattan   
27   20K609         Urban Assembly School for Criminal Justice       Brooklyn   
28   28Q284                         York Early College Academy         Queens   
29   21K348                   High School of Sports Management       Brooklyn   
..      ...                                                ...            ...   
405  24Q264                  Academy of Finance and Enterprise         Queens   
406  08X519  Felisa Rincon de Gautier Institute for Law and...          Bronx   
407  02M374                          Gramercy Arts High School      Manhattan   
408  17K533                School for Democracy and Leadership       Brooklyn   
409  26Q415                    Benjamin N. Cardozo High School         Queens   
410  31R455                            Tottenville High School  Staten Island   
411  25Q263                 Flushing International High School         Queens   
412  02M449                               Vanguard High School      Manhattan   
413  32K552                          Academy of Urban Planning       Brooklyn   
414  06M468        High School for Health Careers and Sciences      Manhattan   
415  30Q580          Baccalaureate School for Global Education         Queens   
416  12X684                                      Wings Academy          Bronx   
417  22K405                                Midwood High School       Brooklyn   
418  25Q670            Robert F. Kennedy Community High School         Queens   
419  19K618                   Academy of Innovative Technology       Brooklyn   
420  02M298                                   Pace High School      Manhattan   
421  08X376  Antonia Pantoja Preparatory Academy: A College...          Bronx   
422  02M500                Unity Center

In [69]:
columns = ['lat', 'lon']
for c in columns:
    data["hsdirectory"][c] = data["hsdirectory"][c].apply(pd.to_numeric, errors='ignore')

In [72]:
data["hsdirectory"]

DBN                                        school_name           boro  \
0    17K548                Brooklyn School for Music & Theatre       Brooklyn   
1    09X543                   High School for Violin and Dance          Bronx   
2    09X327        Comprehensive Model School Project M.S. 327          Bronx   
3    02M280     Manhattan Early College School for Advertising      Manhattan   
4    28Q680  Queens Gateway to Health Sciences Secondary Sc...         Queens   
5    08X348              Schuylerville Preparatory High School          Bronx   
6    14K474      PROGRESS High School for Professional Careers       Brooklyn   
7    02M420  High School for Health Professions and Human S...      Manhattan   
8    28Q350                    Jamaica Gateway to the Sciences         Queens   
9    29Q313  Benjamin Franklin High School for Finance & In...         Queens   
10   12X271                  East Bronx Academy for the Future          Bronx   
11   02M520   Murry Bergtraum High School for Business Careers      Manhattan   
12   06M211  Inwood Early College for Health and Informatio...      Manhattan   
13   24Q600        Queens Vocational and Technical High School         Queens   
14   10X368           In-Tech Academy (M.S. / High School 368)          Bronx   
15   06M348   Washington Heights Expeditionary Learning School      Manhattan   
16   18K629         Cultural Academy for the Arts and Sciences       Brooklyn   
17   03M299      High School for Arts, Imagination and Inquiry      Manhattan   
18   07X520         Foreign Language Academy of Global Studies          Bronx   
19   08X312                             Millennium Art Academy          Bronx   
20   08X293  Renaissance High School for Musical Theater & ...          Bronx   
21   13K419  Science Skills Center High School for Science,...       Brooklyn   
22   31R460                        Susan E. Wagner High School  Staten Island   
23   28Q328               High School for Community Leadership         Queens   
24   05M362                          Columbia Secondary School      Manhattan   
25   10X243                  West Bronx Academy for the Future          Bronx   
26   03M307       Urban Assembly School for Media Studies, The      Manhattan   
27   20K609         Urban Assembly School for Criminal Justice       Brooklyn   
28   28Q284                         York Early College Academy         Queens   
29   21K348                   High School of Sports Management       Brooklyn   
..      ...                                                ...            ...   
405  24Q264                  Academy of Finance and Enterprise         Queens   
406  08X519  Felisa Rincon de Gautier Institute for Law and...          Bronx   
407  02M374                          Gramercy Arts High School      Manhattan   
408  17K533                School for Democracy and Leadership       Brooklyn   
409  26Q415                    Benjamin N. Cardozo High School         Queens   
410  31R455                            Tottenville High School  Staten Island   
411  25Q263                 Flushing International High School         Queens   
412  02M449                               Vanguard High School      Manhattan   
413  32K552                          Academy of Urban Planning       Brooklyn   
414  06M468        High School for Health Careers and Sciences      Manhattan   
415  30Q580          Baccalaureate School for Global Education         Queens   
416  12X684                                      Wings Academy          Bronx   
417  22K405                                Midwood High School       Brooklyn   
418  25Q670            Robert F. Kennedy Community High School         Queens   
419  19K618                   Academy of Innovative Technology       Brooklyn   
420  02M298                                   Pace High School      Manhattan   
421  08X376  Antonia Pantoja Preparatory Academy: A College...          Bronx   
422  02M500                Unity Center

It worked!

Check all the data:

In [73]:
for k,v in data.items():
    print(k)
    print(v.head())

Math_test_results
        DBN Grade  Year      Category  Number Tested Mean Scale Score  \
111  01M034     8  2011  All Students             48              646   
280  01M140     8  2011  All Students             61              665   
346  01M184     8  2011  All Students             49              727   
388  01M188     8  2011  All Students             49              658   
411  01M292     8  2011  All Students             49              650   

    Level 1 # Level 1 % Level 2 # Level 2 % Level 3 # Level 3 % Level 4 #  \
111        15     31.3%        22     45.8%        11     22.9%         0   
280         1      1.6%        43     70.5%        17     27.9%         0   
346         0        0%         0        0%         5     10.2%        44   
388        10     20.4%        26     53.1%        10     20.4%         3   
411        15     30.6%        25       51%         7     14.3%         2   

    Level 4 % Level 3+4 # Level 3+4 %  
111        0%          11       22.9%  
